commands/utils/quick-pick-azure.ts

```diff
@@ -0,0 +1,168 @@
+    let desiredSku: IAzureQuickPickItem<string> = await ext.ui.showQuickPick(quickPickSkuList, { 'canPickMany': false, 'placeHolder': 'Choose a SKU to use' });
+    return desiredSku.data;
+}
+
+export async function quickPickSubscription(): Promise<Subscription> {
+    const subscriptions = AzureUtilityManager.getInstance().getFilteredSubscriptionList();
+    if (subscriptions.length === 0) {
+        vscode.window.showErrorMessage("You do not have any subscriptions. You can create one in your Azure portal", "Open Portal").then(val => {
+            if (val === "Open Portal") {
+                opn('https://portal.azure.com/');
+            }
+        });
+        throw new Error('User has no azure subscriptions');
+    }
+    let quickPickSubList = subscriptions.map(sub => <IAzureQuickPickItem<Subscription>>{ label: sub.displayName, data: sub });
+    let desiredSub = await ext.ui.showQuickPick(quickPickSubList, {
+        'canPickMany': false,
+        'placeHolder': 'Choose a subscription to use'
+    });
+    return desiredSub.data;
+}
+
+export async function quickPickLocation(subscription: Subscription): Promise<string> {
+    let locations: Location[] = await AzureUtilityManager.getInstance().getLocationsBySubscription(subscription);
+    let quickPickLocList = locations.map(loc => <IAzureQuickPickItem<Location>>{ label: loc.displayName, data: loc });
+
+    quickPickLocList.sort((loc1, loc2): number => {
+        return loc1.data.displayName.localeCompare(loc2.data.displayName);
+    });
+
+    let desiredLocation: IAzureQuickPickItem<Location> = await ext.ui.showQuickPick(quickPickLocList, {
+        'canPickMany': false,
+        'placeHolder': 'Choose a location to use'
+    });
+    return desiredLocation.label;
+}
+
+export async function quickPickResourceGroup(canCreateNew?: boolean, subscription?: Subscription): Promise<ResourceGroup> {
+    let resourceGroups = await AzureUtilityManager.getInstance().getResourceGroups(subscription);
+    let resourceGroupNames: string[] = [];
+
+    if (canCreateNew) { resourceGroupNames.push('+ Create new resource group'); }
+    for (let resGroupName of resourceGroups) {
+        resourceGroupNames.push(resGroupName.name);
+    }
+
+    let resourceGroupName = await vscode.window.showQuickPick(resourceGroupNames, {
+        'canPickMany': false,
+        'placeHolder': 'Choose a resource group to use'
+    });
+
+    let resourceGroup: ResourceGroup;
+    if (canCreateNew && resourceGroupName === resourceGroupNames[0]) {
+        if (!subscription) {
+            subscription = await quickPickSubscription();
+        }
+        const loc = await quickPickLocation(subscription);
+        resourceGroup = await createNewResourceGroup(loc, subscription);
+    } else {
+        resourceGroup = resourceGroups.find(resGroup => { return resGroup.name === resourceGroupName; });
+    }
+    return resourceGroup;
```

Comment:
This method has some possible bug points that result from that fact that:
1) You're treating the first item as special and indexing resourceGroupNames[0] directly
2) You're assuming you will find the resourceGroup by name

These are fine as currently written, but they're fragile.  E.g., what would you theoretically do you do if the find returns undefined?  You can avoid these issues completely by rewriting this way:

export async function quickPickResourceGroup(canCreateNew?: boolean, subscription?: Subscription): Promise<ResourceGroup> {
    let resourceGroups = await AzureUtilityManager.getInstance().getResourceGroups(subscription);
    let resourceGroupItems: QuickPickItemWithData<ResourceGroup | undefined>[] = [];
    let createNewItem: QuickPickItemWithData<ResourceGroup | undefined> = { label: '+ Create new resource group', data: undefined };

    if (canCreateNew) { resourceGroupItems.push(createNewItem); }
    for (let resGroup of resourceGroups) {
        resourceGroupItems.push({ label: resGroup.name, data: resGroup });
    }

    let selectedItem = await vscode.window.showQuickPick(resourceGroupItems, {
        'canPickMany': false,
        'placeHolder': 'Choose a resource group to use'
    });

    let resourceGroup: ResourceGroup;
    if (selectedItem === createNewItem) {
        if (!subscription) {
            subscription = await quickPickSubscription();
        }
        const loc = await quickPickLocation(subscription);
        resourceGroup = await createNewResourceGroup(loc, subscription);
    } else {
        resourceGroup = selectedItem.data;
    }

    return resourceGroup;
}

Reply:
Thanks! that looks a lot better and is more stable, I also implemented it for the optional can create part of quickPickACRRegistry. As a note I used IAzureQuickPickItem instead of QuickPickWithData as the IAzure one keeps track of recently used items.